rmw_dds_common/include/rmw_dds_common/qos.hpp

```diff
@@ -56,6 +59,136 @@ qos_profile_check_compatible(
+ * profile such that is matches all subscriptions while maintaining the highest level of service.
+ *
+ * If reliability is BEST_AVAILABLE, then it will be set to RELIABLE.
+ *
+ * If durability is BEST_AVAILABLE, then it will be set to TRANSIENT_LOCAL.
+ *
+ * If liveliness is BEST_AVAILABLE, then it will be set to MANUAL_BY_TOPIC if at least one
+ * subscription is MANUAL_BY_TOPIC.
+ * Otherwise, liveliness will be set to AUTOMATIC.
+ *
+ * If deadline is BEST_AVAILABLE, then it will be set to DEFAULT if all subscriptions are DEFAULT.
+ * Otherwise, deadline will be set to the minimum deadline of all subscriptions.
+ *
+ * If liveliness lease duration is BEST_AVAILABLE, then it will be set to DEFAULT if all
+ * subscriptions are DEFAULT.
+ * Otherwise, liveliness lease duration will be set to the mininum deadline of all subscriptions.
+ *
+ * History, history depth, and lifespan policies are not changed by this function.
+ *
+ * \param[in] subscriptions_info: Endpoint information for subscriptions.
+ * \param[out] publisher_profile: QoS profile that is compatible with the majority of
+ *   the input subscriptions.
+ * \return `RMW_RET_OK` if the operation was successful, or
+ * \return `RMW_RET_INVALID_ARGUMENT` if `subscriptions_info` is `nullptr`, or
+ * \return `RMW_RET_INVALID_ARGUMENT` if `publisher_profile` is `nullptr`, or
+ * \return `RMW_RET_ERROR` if there is an unexpected error.
+ */
+RMW_DDS_COMMON_PUBLIC
+rmw_ret_t
+qos_profile_get_best_available_for_publisher(
+  const rmw_topic_endpoint_info_array_t * subscriptions_info,
+  rmw_qos_profile_t * publisher_profile);
+
+/// Signature matching rmw_get_publishers_info_by_topic and rmw_get_subscriptions_info_by_topic
+using GetEndpointInfoByTopicFunction = std::function<rmw_ret_t(
+      const rmw_node_t *,
+      rcutils_allocator_t *,
+      const char *,
+      bool,
+      rmw_topic_endpoint_info_array_t *)>;
+
+/// Update a subscription QoS profile so that it is compatible with discovered publishers.
+/**
+ * If any policies in `qos_profile` are set to BEST_AVAILABLE, then `get_endpoint_info` will
+ * be called to query endpoint information for adapting the QoS policies.
+ *
+ * For rules related to adapting 'best available' policies, see
+ * \ref `qos_profile_get_best_available_for_subscription`.
+ *
+ * \param[in] node: Node used to query the graph.
+ * \param[in] topic_name: Query info for publishers on this topic name.
+ * \param[inout] qos_profile: Any policies that are set to 'best available' will by updated based
+ *   on publisher endpoint QoS policies.
+ * \param[in] get_endpoint_info: The function used to query for publisher endpoint information.
+ *   I.e. an implementation of `rmw_get_publishers_info_by_topic`.
+ */
+RMW_DDS_COMMON_PUBLIC
+rmw_ret_t
+qos_profile_get_best_available_for_topic_subscription(
+  const rmw_node_t * node,
+  const char * topic_name,
+  rmw_qos_profile_t * qos_profile,
+  GetEndpointInfoByTopicFunction get_endpoint_info);
```

Comment:
can this be a const?

Reply:
Made it a const reference: [efcf049](https://github.com/ros2/rmw_dds_common/pull/60/commits/efcf04925bb228a5ac8e0a8ac92998392f075219)